# Knihovna pandas a základní manipulace s tabulkovými daty

Před jakoukoli složitější analýzou je třeba naučit se základy práce se zpracovávanými daty. Data můžou být různého charakteru - jednorozměrná, dvourozměrná, strukturovaná, nestrukturovaná, obrazová, zvuková... V lekcích expolarativní datové analýzy budeme pracovat s daty převážně **tabulkovými** - tedy takovými, které jistě znáš ze svého oblíbeného (nebo neoblíbeného) tabulkového procesoru (spreadsheetu). Obvykle každý **řádek** takové tabulky odpovídá nějaké věci, exempláři čehosi, případně nějakému pozorování. V jednotlivých **sloupcích** se pak nacházejí jednotlivé vlastnosti či měřené veličiny pro tyto věci charakteristické.

Ve světě Pythonu je pro zpracování tabulkových dat nejpoužívanější knihovnou **pandas**. Ta umožňuje načítat data z mnohých formátů (včetně sešitů XLS(x)), různě je upravovat, velice efektivně počítat se sloupci, přímo zkoumat některé statistické ukazatele a v neposlední řadě výsledky pěkně vizualizovat. Tato lekce tě seznámí se základními použitými koncepty a prací se sloupci.

Více o knihovně pandas najdeš na jejích domovských stránkách: https://pandas.pydata.org/

## Import knihoven

In [1]:
import pandas as pd     # K pandas budeme přistupovat pomocí aliasu pd

Při běžném programování se snažíme vyhnout aliasům, protože znepřehledňují kód, ale ve světě datové analytiky se některé aliasy tak zažily, že je budeme používat i v tomto kurzu (další ještě přijdou).

## Načtení tabulky s daty

Skočíme do pandas rovnýma nohama a ukážeme si typický příklad dat, která s touto knihovnou budeme zpracovávat.

Pro čtení dat má Pandas celou řadu funkcí `read_*`, díky kterým si poradí s celou řadou různých formátů. Poměrně častým je formát CSV (comma-separated values - [wiki](https://cs.wikipedia.org/wiki/CSV)), ve kterém jsou jednotlivé hodnoty oddělené čárkami.

**TODO** Ověřit následující: mechanismus

Pokud budeš tímto notebookem experimentovat, stáhni si nejdříve soubor s daty z [tohoto odkazu](static/pokemon.csv). Data pro pokusy jsou vytvořena z [komplexního Pokedexu na Githubu](https://github.com/veekun/pokedex).

In [2]:
data = pd.read_csv("static/pokemon.csv")

Data (ať už to je cokoliv) jsou nyní načtená v paměti. Pojďme se podívat, co se v nich ukrývá.

In [3]:
data

id         name  height  weight   color      shape  is baby    type 1  \
0      1    bulbasaur     0.7     6.9   green  quadruped    False     Grass   
1      2      ivysaur     1.0    13.0   green  quadruped    False     Grass   
2      3     venusaur     2.0   100.0   green  quadruped    False     Grass   
3      4   charmander     0.6     8.5     red    upright    False      Fire   
4      5   charmeleon     1.1    19.0     red    upright    False      Fire   
..   ...          ...     ...     ...     ...        ...      ...       ...   
802  803      poipole     0.6     1.8  purple    upright    False    Poison   
803  804    naganadel     3.6   150.0  purple      wings    False    Poison   
804  805    stakataka     5.5   820.0    gray  quadruped    False      Rock   
805  806  blacephalon     1.8    13.0   white   humanoid    False      Fire   
806  807      zeraora     1.5    44.5  yellow   humanoid    False  Electric   

     type 2  hp  attack  defense  speed  
0    Poison  45      49       49     45  
1    Poison  60      62       63     60  
2    Poison  80      82       83     80  
3       NaN  39      52       43     65  
4       NaN  58      64       58     80  
..      ...  ..     ...      ...    ...  
802     NaN  67      73       67     73  
803  Dragon  73      73       73    121  
804   Steel  61     131      211     13  
805   Ghost  53     127       53    107  
806     NaN  88     112       75    143  

[807 rows x 13 columns]

Pokud vše zafungovalo, jak má, měl(a) bys před sebou mít relativně pěkně naformátovanou tabulku. Základní zobrazení v notebooku nám ukáže prvních pět a posledních pět řádků (kdo by riskoval, že mu tisíce řádků zahltí okno prohlížeče?) v tabulce společně s informací o celkovém počtu řádků a sloupců. V tomto případě tabulka obsahuje celkem 13 vlastností (pojmenovaných sloupců) 807 různých pokémonů (očíslovaných řádků).

**Poznámka:** V tomto jednoduchém případě se tabulka načetla správně hned na první pokus, bez jakýchkoliv specifických parameterů, všechny sloupce se zdají obsahovat použitelné hodnoty. To je (zejména u formátu CSV) vlastně z pekla štěstí. Obvykle jsou se vstupními daty problémy - kupř. nemají popsané sloupce (případně je mají popsané podivně), používají zvláštní oddělovače záznamů nebo desetinné části čísel, mnohé řádky obsahují chybějící (nebo špatně opsané) hodnoty, ... **Čištění dat** se budeme věnovat někdy příště.

Co je ten objekt uložený v proměnné `data` vlastně zač? Jaké je třídy?

In [4]:
type(data)

pandas.core.frame.DataFrame

## DataFrame a Series

Odpověď zní `DataFrame`. Tento termín (používaný i v dalších oblíbených statistických jazycích, například "R") je nespíš bez českého ekvivalentu, a tak nadále budeme mluvit o tabulkách či o instancích třídy DataFrame. Pokusme se nyní náš první DataFrame rozpitvat.



In [5]:
data["name"]

0        bulbasaur
1          ivysaur
2         venusaur
3       charmander
4       charmeleon
          ...     
802        poipole
803      naganadel
804      stakataka
805    blacephalon
806        zeraora
Name: name, Length: 807, dtype: object

## Sloupce a řádky

Sloupce i řádky mají tzv. index, s jehož pomocí se dá k jednotlivým částem tabulky přistoupit. V našem případě má každý řádek automaticky doplněný unikátní číselný index (ten bezejmenný sloupec nalevo) a jako index pro sloupce nám poslouží prvni řádek CSV souboru s názvy sloupců.

Později se naučíme idexy měnit a vkládat do nich i více než jednu hodnotu.

### Sloupce

Sloupce lze vybrat stejně jako se vybírají hodnoty ze slovníku - do hranatých závorek stačí napsat název sloupce.

In [6]:
data['name']

0        bulbasaur
1          ivysaur
2         venusaur
3       charmander
4       charmeleon
          ...     
802        poipole
803      naganadel
804      stakataka
805    blacephalon
806        zeraora
Name: name, Length: 807, dtype: object

Pokud potřebujeme více sloupců najednou, vložíme do hranatých závorek za proměnnou s tabulkou jejich seznam.

In [7]:
data[['name', 'color']]

name   color
0      bulbasaur   green
1        ivysaur   green
2       venusaur   green
3     charmander     red
4     charmeleon     red
..           ...     ...
802      poipole  purple
803    naganadel  purple
804    stakataka    gray
805  blacephalon   white
806      zeraora  yellow

[807 rows x 2 columns]

Pokud název sloupce neobsahuje mezeru a nekoliduje s názvem existující metody, dá se k němu přistoupit i pomocí tečkové notace, což je velmi užitečné především při psaní složitějších podmínek.

In [8]:
data.name

0        bulbasaur
1          ivysaur
2         venusaur
3       charmander
4       charmeleon
          ...     
802        poipole
803      naganadel
804      stakataka
805    blacephalon
806        zeraora
Name: name, Length: 807, dtype: object

Pro sloupec `shape` je tento přístup nepoužitelný, protože koliduje s atributem, ve kterém je uložen počet řádků a slupců tabulky.

In [9]:
data.shape

(807, 13)

### Řádky

Pro řádky se používá tzv. indexer `loc`, který nám dokáže obstarat jeden či více řádků.

Můžeme si například říci o jeden řádek jeho indexem:

In [10]:
data.loc[1]

id                 2
name         ivysaur
height             1
weight            13
color          green
shape      quadruped
is baby        False
type 1         Grass
type 2        Poison
hp                60
attack            62
defense           63
speed             60
Name: 1, dtype: object

O více řádků seznamem jejich indexů:

In [11]:
data.loc[[1, 2, 3]]

id        name  height  weight  color      shape  is baby type 1  type 2  \
1   2     ivysaur     1.0    13.0  green  quadruped    False  Grass  Poison   
2   3    venusaur     2.0   100.0  green  quadruped    False  Grass  Poison   
3   4  charmander     0.6     8.5    red    upright    False   Fire     NaN   

   hp  attack  defense  speed  
1  60      62       63     60  
2  80      82       83     80  
3  39      52       43     65

Nebo třeba stejně jako i seznamů o několik řádků pomocí rozsahu indexů:

In [12]:
data.loc[100:105]

id       name  height  weight   color     shape  is baby    type 1  \
100  101  electrode     1.2    66.6     red      ball    False  Electric   
101  102  exeggcute     0.4     2.5    pink     heads    False     Grass   
102  103  exeggutor     2.0   120.0  yellow      legs    False     Grass   
103  104     cubone     0.4     6.5   brown   upright    False    Ground   
104  105    marowak     1.0    45.0   brown   upright    False    Ground   
105  106  hitmonlee     1.5    49.8   brown  humanoid    False  Fighting   

      type 2  hp  attack  defense  speed  
100      NaN  60      50       70    150  
101  Psychic  60      40       80     40  
102  Psychic  95      95       85     55  
103      NaN  50      50       95     35  
104      NaN  60      80      110     45  
105      NaN  50     120       53     87

### Konkrétní hodnoty

Indexer `loc` je vhodný i pro získání konkrétní hodnoty, jen se při indexaci použijí oba indexy (pro řádek i sloupec) najednou ve formě ntice.

In [13]:
data.loc[101, 'name']

'exeggcute'

## Filtrace dat

Velmi často se hodí získat část tabulky, ve kteté hodnoty splňují určitá kritéria. Toho lze docílit tak, že napíšeme do hranatých závorek podmínku a dostaneme jen ty řádky, které danou podmínku splňují.

Například nás budou zajímat jen červení pokémoni.

In [14]:
data[data['color'] == "red"]

id        name  height  weight color      shape  is baby type 1  \
3      4  charmander     0.6     8.5   red    upright    False   Fire   
4      5  charmeleon     1.1    19.0   red    upright    False   Fire   
5      6   charizard     1.7    90.5   red    upright    False   Fire   
44    45   vileplume     1.2    18.6   red   humanoid    False  Grass   
45    46       paras     0.3     5.4   red      armor    False    Bug   
..   ...         ...     ...     ...   ...        ...      ...    ...   
726  727  incineroar     1.8    83.0   red    upright    False   Fire   
740  741    oricorio     0.6     3.4   red      wings    False   Fire   
775  776  turtonator     2.0   212.0   red    upright    False   Fire   
786  787   tapu-bulu     1.9    45.5   red       arms    False  Grass   
793  794    buzzwole     2.4   333.6   red  tentacles    False    Bug   

       type 2   hp  attack  defense  speed  
3         NaN   39      52       43     65  
4         NaN   58      64       58     80  
5      Flying   78      84       78    100  
44     Poison   75      80       85     50  
45      Grass   35      70       55     25  
..        ...  ...     ...      ...    ...  
726      Dark   95     115       90     60  
740    Flying   75      70       70     93  
775    Dragon   60      78      135     36  
786     Fairy   70     130      115     75  
793  Fighting  107     139      139     79  

[82 rows x 13 columns]

Podmínky se dají i kombinovat, ale musí se dodržet jistá pravidla. První z nich je, že každá podmínka musí být samostatně uzavřena v kulatých závorkách. Dále pak logické spojky `and` a `or` jsou zde nahrazeny znaky `&` a `|`.

Co třeba červení pokémoni, jejichž síla útoku je vyšší nebo rovna 130?

In [15]:
data[(data['attack'] >= 130) & (data['color'] == "red")]

id        name  height  weight color      shape  is baby   type 1  \
98    99     kingler     1.3    60.0   red      armor    False    Water   
135  136     flareon     0.9    25.0   red  quadruped    False     Fire   
211  212      scizor     1.8   118.0   red  bug-wings    False      Bug   
249  250       ho-oh     3.8   199.0   red      wings    False     Fire   
382  383     groudon     3.5   950.0   red    upright    False   Ground   
385  386      deoxys     1.7    60.8   red   humanoid    False  Psychic   
554  555  darmanitan     1.3    92.9   red  quadruped    False     Fire   
716  717     yveltal     5.8   203.0   red      wings    False     Dark   
786  787   tapu-bulu     1.9    45.5   red       arms    False    Grass   
793  794    buzzwole     2.4   333.6   red  tentacles    False      Bug   

       type 2   hp  attack  defense  speed  
98        NaN   55     130      115     75  
135       NaN   65     130       60     65  
211     Steel   70     130      100     65  
249    Flying  106     130       90     90  
382       NaN  100     150      140     90  
385       NaN   50     150       50    150  
554       NaN  105     140       55     95  
716    Flying  126     131       95     99  
786     Fairy   70     130      115     75  
793  Fighting  107     139      139     79

## Operace se sloupci

Podmínka zapsaná v hranatých závorkách ve skutečnosti není jen tak obyčejnou podmínkou. Jedná se totiž o hromadnou operaci napříč celým sloupcem, která pro každý řádek vrátí `True` nebo `False`, což v případě použití k filtraci rozhodne o jeho zobrazení.

In [16]:
data.speed >= 65

0      False
1      False
2       True
3       True
4       True
       ...  
802     True
803     True
804    False
805     True
806     True
Name: speed, Length: 807, dtype: bool

I další známé operátory lze použít stejným způsobem a aplikovat je na sloupec a jednu hodnotu či na více sloupců najednou.

In [17]:
data["speed"] / 10

0       4.5
1       6.0
2       8.0
3       6.5
4       8.0
       ... 
802     7.3
803    12.1
804     1.3
805    10.7
806    14.3
Name: speed, Length: 807, dtype: float64

In [18]:
data.hp * data.defense

0       2205
1       3780
2       6640
3       1677
4       3364
       ...  
802     4489
803     5329
804    12871
805     2809
806     6600
Length: 807, dtype: int64

Výsledek takové operace lze velmi snadno přidat zpět k původní tabulce jako nový sloupec.

In [19]:
data["fast"] = data.speed >= 65

In [20]:
data

id         name  height  weight   color      shape  is baby    type 1  \
0      1    bulbasaur     0.7     6.9   green  quadruped    False     Grass   
1      2      ivysaur     1.0    13.0   green  quadruped    False     Grass   
2      3     venusaur     2.0   100.0   green  quadruped    False     Grass   
3      4   charmander     0.6     8.5     red    upright    False      Fire   
4      5   charmeleon     1.1    19.0     red    upright    False      Fire   
..   ...          ...     ...     ...     ...        ...      ...       ...   
802  803      poipole     0.6     1.8  purple    upright    False    Poison   
803  804    naganadel     3.6   150.0  purple      wings    False    Poison   
804  805    stakataka     5.5   820.0    gray  quadruped    False      Rock   
805  806  blacephalon     1.8    13.0   white   humanoid    False      Fire   
806  807      zeraora     1.5    44.5  yellow   humanoid    False  Electric   

     type 2  hp  attack  defense  speed   fast  
0    Poison  45      49       49     45  False  
1    Poison  60      62       63     60  False  
2    Poison  80      82       83     80   True  
3       NaN  39      52       43     65   True  
4       NaN  58      64       58     80   True  
..      ...  ..     ...      ...    ...    ...  
802     NaN  67      73       67     73   True  
803  Dragon  73      73       73    121   True  
804   Steel  61     131      211     13  False  
805   Ghost  53     127       53    107   True  
806     NaN  88     112       75    143   True  

[807 rows x 14 columns]

## Řazení

K řazení podle hodnot slouží metoda `sort_values`, podle indexu pak řadí metoda `sort_index`. Prvních několik řádků získáme metodou `head` a naopak poslední nám poskytne metoda `tail`.

Pět nejrychlejších pokémonů tedy získáme snadno sestupným seřazením podle jejich rychlosti a omezením na prvních pět řádků.

In [21]:
data.sort_values(by="speed", ascending=False).head()

id       name  height  weight   color      shape  is baby    type 1  \
290  291    ninjask     0.8    12.0  yellow  bug-wings    False       Bug   
794  795  pheromosa     1.8    25.0   white   humanoid    False       Bug   
385  386     deoxys     1.7    60.8     red   humanoid    False   Psychic   
100  101  electrode     1.2    66.6     red       ball    False  Electric   
616  617   accelgor     0.8    25.3     red       arms    False       Bug   

       type 2  hp  attack  defense  speed  fast  
290    Flying  61      90       45    160  True  
794  Fighting  71     137       37    151  True  
385       NaN  50     150       50    150  True  
100       NaN  60      50       70    150  True  
616       NaN  80      70       40    145  True